lib/common/tests/rules/pcmk__cmp_by_type_test.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2024 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+
+#include <crm/common/rules_internal.h>
+#include <crm/common/unittest_internal.h>
+
+static void
+null_compares_lesser(void **state)
+{
+    assert_int_equal(pcmk__cmp_by_type(NULL, NULL, pcmk__type_string), 0);
+    assert_true(pcmk__cmp_by_type("0", NULL, pcmk__type_integer) > 0);
+    assert_true(pcmk__cmp_by_type(NULL, "0", pcmk__type_number) < 0);
+}
+
+static void
+invalid_compares_equal(void **state)
+{
+    assert_int_equal(pcmk__cmp_by_type("0", "1", pcmk__type_unknown), 0);
+    assert_int_equal(pcmk__cmp_by_type("hi", "bye", pcmk__type_unknown), 0);
+    assert_int_equal(pcmk__cmp_by_type("-1.0", "2.0", pcmk__type_unknown), 0);
+}
+
+static void
+compare_string_type(void **state)
+{
+    assert_int_equal(pcmk__cmp_by_type("bye", "bye", pcmk__type_string), 0);
+    assert_int_equal(pcmk__cmp_by_type("bye", "BYE", pcmk__type_string), 0);
+    assert_true(pcmk__cmp_by_type("bye", "hello", pcmk__type_string) < 0);
+    assert_true(pcmk__cmp_by_type("bye", "HELLO", pcmk__type_string) < 0);
+    assert_true(pcmk__cmp_by_type("bye", "boo", pcmk__type_string) > 0);
+    assert_true(pcmk__cmp_by_type("bye", "Boo", pcmk__type_string) > 0);
+}
+
+static void
+compare_integer_type(void **state)
+{
+    assert_int_equal(pcmk__cmp_by_type("0", "0", pcmk__type_integer), 0);
+    assert_true(pcmk__cmp_by_type("0", "1", pcmk__type_integer) < 0);
+    assert_true(pcmk__cmp_by_type("1", "0", pcmk__type_integer) > 0);
+    assert_true(pcmk__cmp_by_type("3999", "399", pcmk__type_integer) > 0);
+
+    // Non-integers compare as strings
+    assert_true(pcmk__cmp_by_type("0", "x", pcmk__type_integer) < 0);
```

Comment:
or could be explicit in the code for these "compare as strings" tests
```
assert_int_equal(pcmk__cmp_by_type("0", "x", pcmk__type_integer),
                 pcmk__cmp_by_type("0", "x", pcmk__type_string));

Reply:
good idea